### PR TITLE
Feature/add vectorised scoring scheme

### LINKED ISSUE #1319: ### Vectorise Scoring Schemes - level1
```c++
using score_type = int;
using simd_t = simd_type<score_type>;

nucleotide_scoring_scheme nucl_scheme{};

vectorised_scoring_scheme<simd_t> vec_scheme{nucl_scheme};

simd_t a = seqan3::simd::fill<simd_t>(1); // [C,C,C,C,C,C,...]
simd_t b = seqan3::simd::fill<simd_t>(2);
b[0] = 1; // [C,G,G,G,G,G,...]
vec_scheme.score(a, b); // [0, -1, -1, -1, ...]
```

**TASK: write class that converts a scoring_scheme into a vectorised scoring_scheme I/II**

- [x] TODO: where to place this class
- [x] Add `vectorised_scoring_scheme` in [include/seqan3/alignment/scoring](https://github.com/seqan/seqan3/tree/8993bb63efa8f844f23bd2207e9bac1061c94bd1/include/seqan3/alignment/scoring)
- [x] Write tests, s.t. vectorised_scoring_scheme will be tested on any given scoring scheme
- [x] (NOTE: You don't have to vectorise anything, or better stated you don't have to do anything clever, you can loop in the score function and produce the result)

### LINKED ISSUE #1333: ### Vectorise Scoring Schemes - level1
**TASK: write class that converts a scoring_scheme into a vectorised scoring_scheme II/II**

- [x] `seqan3::simd::simd_traits<simd_t>::length` might be helpful; see [simd_traits.hpp](https://github.com/seqan/seqan3/blob/8993bb63efa8f844f23bd2207e9bac1061c94bd1/include/seqan3/core/simd/simd_traits.hpp#L37-L64)
- [x] (NOTE: this class does not need to fulfil the [scoring_scheme concept](https://github.com/seqan/seqan3/blob/8993bb63efa8f844f23bd2207e9bac1061c94bd1/include/seqan3/alignment/scoring/scoring_scheme_concept.hpp#L60-L68)

### PR DESCRIPTION
**Pull request by rrahn**
Resolves #1319, 
Resolves #1333

Adds vectorised scoring scheme for simple scores differentiating between match and mismatch scores only. The vectorised version lives in detail as it is not exposed to the user.

### CODE REVIEW

**Comment by rrahn** [comment]
@smehringer ping

File: test/unit/alignment/scoring/simd_match_mismatch_scoring_scheme_test.cpp [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    simd_value2 = simd::fill<TypeParam>(1);
+    SIMD_EQ(simd_scheme.score(simd_value1, simd_value2), simd::fill<TypeParam>(-5));
+}
+
+TYPED_TEST(simd_match_mismatch_scoring_scheme_test, construct_from_scoring_scheme_throw_on_overflow)
+{
+    using scalar_t = typename simd_traits<TypeParam>::scalar_type;
+    using scheme_t = simd_match_mismatch_scoring_scheme<TypeParam, dna4, detail::global_alignment_type>;
+
+    int64_t to_big = static_cast<int64_t>(std::numeric_limits<scalar_t>::max()) + 1;
```
Comment (joshuak94) [inline test/unit/alignment/scoring/simd_match_mismatch_scoring_scheme_test.cpp]:
```suggestion
    int64_t too_big = static_cast<int64_t>(std::numeric_limits<scalar_t>::max()) + 1;
```

File: test/unit/alignment/scoring/simd_match_mismatch_scoring_scheme_test.cpp [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    int64_t to_small = static_cast<int64_t>(std::numeric_limits<scalar_t>::lowest()) - 1;
```
Comment (joshuak94) [inline test/unit/alignment/scoring/simd_match_mismatch_scoring_scheme_test.cpp]:
```suggestion
    int64_t too_small = static_cast<int64_t>(std::numeric_limits<scalar_t>::lowest()) - 1;
```

File: include/seqan3/alignment/scoring/detail/simd_match_mismatch_scoring_scheme.hpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+ *
+ * During the vectorised alignment multiple sequences are packed into one simd vector.
+ * To handle sequences with different sizes in the vectorised alignment algorithm the smaller sequences are filled up
+ * with special padding symbols. These padding symbols are chosen in a way that allow the computation of the alignments
+ * without the need of masking the results for invalid positions within the matrix because a specific position might
+ * have exceeded the original sequence size.
+ * To do so, the global alignment uses the same padding symbol for the first sequence pack and the second sequence
+ * pack. This padding symbol is distinct to any symbol in the underlying alphabet of the sequences.
+ * The score function is adapted in a way that a comparison with a padding symbol always yields a match.
+ * Thus, after the end of a sequence within the pack has reached the score can only grow.
```
Comment (joshuak94) [inline include/seqan3/alignment/scoring/detail/simd_match_mismatch_scoring_scheme.hpp]:
```suggestion
 * Thus, after the end of a sequence within the pack is reached the score can only grow.
```

File: include/seqan3/alignment/scoring/detail/simd_match_mismatch_scoring_scheme.hpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    }
+    //!\}
+
+    /*!\name Score computation
+     * \{
+     */
+    /*!\brief Computes the score for two simd vectors.
+     * \param[in] lhs The left operand to compare.
+     * \param[in] rhs The right operand to compare.
+     * \returns
```
Comment (joshuak94) [inline include/seqan3/alignment/scoring/detail/simd_match_mismatch_scoring_scheme.hpp]: Should there be a description of what is returned?

File: include/seqan3/alignment/scoring/detail/simd_match_mismatch_scoring_scheme.hpp [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+#include <seqan3/alphabet/concept.hpp>
+#include <seqan3/core/concept/cereal.hpp>
+#include <seqan3/core/simd/concept.hpp>
+#include <seqan3/core/simd/simd_algorithm.hpp>
+#include <seqan3/std/concepts>
+
+namespace seqan3::detail
+{
+
+/*!\brief A vectorised scoring scheme scheme handling matches and mismatches only.
```
Comment (joshuak94) [inline include/seqan3/alignment/scoring/detail/simd_match_mismatch_scoring_scheme.hpp]:
```suggestion
/*!\brief A vectorised scoring scheme handling matches and mismatches only.
```

File: test/unit/alignment/scoring/simd_match_mismatch_scoring_scheme_test.cpp (hunk 1 above)
Comment (smehringer) [reply test/unit/alignment/scoring/simd_match_mismatch_scoring_scheme_test.cpp]: not resolved

File: test/unit/alignment/scoring/simd_match_mismatch_scoring_scheme_test.cpp (hunk 2 above)
Comment (smehringer) [reply test/unit/alignment/scoring/simd_match_mismatch_scoring_scheme_test.cpp]: not resolved

File: include/seqan3/alignment/scoring/detail/simd_match_mismatch_scoring_scheme.hpp [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+     * cannot be represented by the scalar type of the used simd vector.
+     */
+    template <typename scoring_scheme_t>
+    constexpr void initialise_from_scalar_scoring_scheme(scoring_scheme_t const & scoring_scheme)
+    {
+        using score_t = decltype(std::declval<scoring_scheme_t const &>().score(seqan3::assign_rank_to(0, alphabet_t{}),
+                                                                                seqan3::assign_rank_to(0, alphabet_t{})));
+        using simd_scalar_t = typename simd_traits<simd_score_t>::scalar_type;
+
+        score_t scalar_match_score = scoring_scheme.score(seqan3::assign_rank_to(0, alphabet_t{}),
```
[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan3/alignment/scoring/detail/simd_match_mismatch_scoring_scheme.hpp]: That is not specified by the alphabet concept is it? 
At least I need to make sure that alphabet{} and assign_rank_to{1, alphabet{}} ist not the same. So using the same expression with a different rank makes at least the intent clear?

File: include/seqan3/alignment/scoring/detail/simd_match_mismatch_scoring_scheme.hpp [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        score_t scalar_mismatch_score = scoring_scheme.score(seqan3::assign_rank_to(0, alphabet_t{}),
+                                                             seqan3::assign_rank_to(1, alphabet_t{}));
+
+        // Check if the scoring scheme match and mismatch scores do not overflow with the respective scalar type.
+        if constexpr (sizeof(simd_scalar_t) < sizeof(score_t))
+        {
+            if (scalar_match_score > static_cast<score_t>(std::numeric_limits<simd_scalar_t>::max()) ||
+                scalar_mismatch_score < static_cast<score_t>(std::numeric_limits<simd_scalar_t>::lowest()))
+            {
+                throw std::invalid_argument{"The selected scoring scheme score overflows "
```
[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan3/alignment/scoring/detail/simd_match_mismatch_scoring_scheme.hpp]: why?

[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan3/alignment/scoring/detail/simd_match_mismatch_scoring_scheme.hpp]: as discussed

**Comment by rrahn** [comment]
> Question: Can an alignment with score_type double be vectorised?

Possibly yes, but not right now as for the moment the simd does not allow floating point as scalar types. But in the future we might add this.

**Pull request merged**